server/src/main/java/org/elasticsearch/common/blobstore/fs/FsBlobContainer.java

```diff
@@ -302,6 +302,32 @@ private void writeToPath(
         IOUtils.fsync(file, false);
     }
 
+    @Override
+    public void writeBlobAtomic(
+        OperationPurpose purpose,
+        String blobName,
+        InputStream inputStream,
+        long blobSize,
+        boolean failIfAlreadyExists
+    ) throws IOException {
+        assert purpose != OperationPurpose.SNAPSHOT_DATA && BlobContainer.assertPurposeConsistency(purpose, blobName) : purpose;
+        final String tempBlob = tempBlobName(blobName);
+        final Path tempBlobPath = path.resolve(tempBlob);
+        try {
+            writeToPath(inputStream, tempBlobPath, blobSize);
```

Comment:
I think we could easily make `writeToPath` return the number of bytes written and then verify that it matches `blobSize`. That would be nice to verify that `blobSize` is filled in correctly.

Reply:
I guess this should be an assertion, right? Added one in writeToPath() itself, and also a similar assertion in the HDFS blob container.

Reply:
Assertion is great, thanks. Not sure if we can apply it in `writePath` though, depends on my other comment.

Reply:
Which `writePath` you mean?

Reply:
Sorry, meant `writeToPath`. Since that is used both for `writeBlob` and `writeBlobAtomic`, we may not be able to assert blob-size is right there, but this all depends on the examination in the other comment.